src/Init/System/IO.lean

```diff
@@ -181,74 +186,72 @@ def fopenFlags (m : FS.Mode) (b : Bool) : String :=
   let bin := if b then "b" else "t"
   mode ++ bin
 
-@[extern "lean_io_prim_handle_mk"] constant Handle.mk (fn : @& FilePath) (mode : @& String) : IO Handle
-@[extern "lean_io_prim_handle_is_eof"] constant Handle.isEof (h : @& Handle) : IO Bool
-@[extern "lean_io_prim_handle_flush"] constant Handle.flush (h : @& Handle) : IO Unit
-@[extern "lean_io_prim_handle_read"] constant Handle.read  (h : @& Handle) (bytes : USize) : IO ByteArray
-@[extern "lean_io_prim_handle_write"] constant Handle.write (h : @& Handle) (buffer : @& ByteArray) : IO Unit
+@[extern "lean_io_prim_handle_mk"] constant mkPrim (fn : @& FilePath) (mode : @& String) : IO Handle
 
-@[extern "lean_io_prim_handle_get_line"] constant Handle.getLine (h : @& Handle) : IO String
-@[extern "lean_io_prim_handle_put_str"] constant Handle.putStr (h : @& Handle) (s : @& String) : IO Unit
+def mk (fn : FilePath) (Mode : Mode) (bin : Bool := true) : IO Handle :=
+  mkPrim fn (fopenFlags Mode bin)
+
+/--
+Returns whether the end of the file has been reached while reading a file.
+`h.isEof` returns true /after/ the first attempt at reading past the end of `h`.
+Once `h.isEof` is true, reading `h` will always return an empty array.
+-/
+@[extern "lean_io_prim_handle_is_eof"] constant isEof (h : @& Handle) : IO Bool
+@[extern "lean_io_prim_handle_flush"] constant flush (h : @& Handle) : IO Unit
+@[extern "lean_io_prim_handle_read"] constant read  (h : @& Handle) (bytes : USize) : IO ByteArray
+@[extern "lean_io_prim_handle_write"] constant write (h : @& Handle) (buffer : @& ByteArray) : IO Unit
+
+@[extern "lean_io_prim_handle_get_line"] constant getLine (h : @& Handle) : IO String
+@[extern "lean_io_prim_handle_put_str"] constant putStr (h : @& Handle) (s : @& String) : IO Unit
+
+end Handle
 
-@[extern "lean_io_getenv"] constant getEnv (var : @& String) : IO (Option String)
 @[extern "lean_io_realpath"] constant realPath (fname : FilePath) : IO FilePath
 @[extern "lean_io_remove_file"] constant removeFile (fname : @& FilePath) : IO Unit
+@[extern "lean_io_create_dir"] constant createDir : @& FilePath → IO Unit
+
+end FS
+
+@[extern "lean_io_getenv"] constant getEnv (var : @& String) : IO (Option String)
 @[extern "lean_io_app_path"] constant appPath : IO FilePath
 @[extern "lean_io_current_dir"] constant currentDir : IO FilePath
 
-end Prim
-
 namespace FS
-variable [Monad m] [MonadLiftT IO m]
-
-def Handle.mk (fn : FilePath) (Mode : Mode) (bin : Bool := true) : m Handle :=
-  liftM (Prim.Handle.mk fn (Prim.fopenFlags Mode bin))
 
 @[inline]
-def withFile (fn : FilePath) (mode : Mode) (f : Handle → m α) : m α :=
+def withFile (fn : FilePath) (mode : Mode) (f : Handle → IO α) : IO α :=
```

Comment:
What is the reasoning for applying this change to functions that take monadic callbacks? I often used `withFile` with an `f` that does not return `IO`. What is the recommended workaround?

Reply:
I thought I had reverted all higher-order uses of `m`, but apparently not. Will revert as soon as the tests pass.

Reply:
Note that this function is a bit weird in that I don't see any real advantage of using it in favor of using `Handle.mk` directly. We don't give any formal guarantees but it should be reasonable to assume that non-borrowed variables are freed (decremented) as soon as possible. And `withFile` doesn't guarantee that the file is closed after executing `f` either since the handle could be passed back in the return type. But perhaps it still gives a kind of peace of mind? @leodemoura Am I missing something?